lib/index.js

```diff
@@ -1,88 +1,356 @@
+};
+
+/**
+ * @summary Checks if the reported issue has been reported previously (duplicated).
+ * @return  {Boolean} Whether the issue is duplicated (`true`) or not (`false`).
+ * @todo    Remove after fixing https://github.com/phpmd/phpmd/issues/467
+ */
+const reportedPreviously = (projectRoot, fingerprint) => {
+  if (!Object.prototype.hasOwnProperty.call(fingerprints, projectRoot)) {
+    fingerprints[projectRoot] = new Set();
+  }
+
+  if (fingerprints[projectRoot].has(fingerprint)) return true;
+
+  fingerprints[projectRoot].add(fingerprint);
+  return false;
+};
+
+/**
+ * Search for a CodeClimate config file in the project tree. If none found,
+ * use the presence of a `.git` directory as the assumed project root.
+ *
+ * @param  {String}  filePath The absolute path to the file triggering the analysis.
+ * @return {Promise}          The absolute path to the project root.
+ */
+const findProjectRoot = async (filePath) => {
+  const fileDir = dirname(filePath);
+  const configurationFilePath = await Helpers.findAsync(fileDir, configurationFile);
+
+  if (configurationFilePath !== null) {
+    return dirname(configurationFilePath);
+  }
+
+  // Fall back to dir of current file if a .git repo can't be found.
+  const gitPath = await Helpers.findAsync(fileDir, '.git');
+  return dirname(gitPath || filePath);
+};
+
+/**
+ * Returns the range (lines/columns) for a given issue from its location.
+ *
+ * @param  {TextEditor} textEditor The Atom TextEditor instance.
+ * @param  {Object}     location   The location object of the CodeClimate issue.
+ * @return {Array[]}               The range: `[[lineNumber, colStart], [lineNumber, colEnd]]`.
+ */
+const calcRange = (textEditor, location) => {
+  // Issue only has a line number
+  if (!Object.prototype.hasOwnProperty.call(location, 'positions')) {
+    return Helpers.generateRange(textEditor, location.lines.begin - 1);
+  }
+
+  const { positions } = location;
+  const line = positions.begin.line - 1;
+
+  // Invalid starting column, just treat it as a line number
+  if (positions.begin.column === undefined) {
+    return Helpers.generateRange(textEditor, line);
+  }
+
+  const colStart = (positions.begin.column - 1) || 0;
+  const colEnd = (positions.end.column === undefined)
+    ? undefined : (positions.end.column - 1);
+
+  // No valid end column, let generateRange highlight a word
+  if (colEnd === undefined || colStart === colEnd) {
+    return Helpers.generateRange(textEditor, line, colStart);
+  }
+
+  // Valid end column, and different from the start one
+  return [[line, colStart], [line, colEnd]];
+};
+
+/**
+ * @summary Fetch the paths of all currently open files on Atom.
+ * @return  {Object} Dictionary of open file abspaths ~> textEditor.
+ *
+ * NOTE Files indexed by abspath to avoid relative filepath collision among different projects.
+ */
+const fetchOpenFilepaths = () => {
+  const openFiles = {};
+  atom.workspace.textEditorRegistry.editors.forEach((textEditor) => {
```

Comment:
`atom.workspace.textEditorRegistry.editors` is a private property, this should be `atom.workspace.getTextEditors()`.

Also instead of calling this every single time, it might be better to do `atom.workspace.observeTextEditors()`.